Issue #7277 - Allow `Request.getLocalName()` and `.getLocalPort()` to be overridden

joakime opened the pull request:
Uses a Connection.Listener to set the override on connections that are opened using an `HttpChannel`.

joakime committed:
Issue #7277 - ServletRequest.getLocalName() override

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Allow Request.getLocalName() to be overridden

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -298,6 +301,26 @@ public int getAcceptors()
+     * @return Returns the optional local name override
+     */
+    @ManagedAttribute("local name override")
+    public String getLocalNameOverride()
```
I think just `getLocalName()` is enough and consistent with all the other APIs of the same name in `HttpChannel` and `Request`.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -318,6 +341,12 @@ protected void doStart() throws Exception
+        if (_localNameOverride != null)
+        {
+            addBean(new HttpChannelLocalNameOverrideListener(_localNameOverride));
+        }
```
Not necessary, see below.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,6 +301,64 @@ public EndPoint getEndPoint()
+    public String getLocalName()
+    {
+        if (overriddenLocalName != null)
+            return overriddenLocalName;
```
From here you can access the `Connector`.

So this code becomes something like:

```java
Connector connector = getConnector();
String localName = null;
if (connector instanceof AbstractConnector)
  localName = ((AbstractConnector)connector).getLocalName();
if (localName != null)
  return localName;
```

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelLocalNameOverrideListener.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Sets the {@link HttpChannel#setLocalNameOverride(String)} on all newly opened connections.
+ */
+public class HttpChannelLocalNameOverrideListener implements Connection.Listener
```
Not necessary.

joakime committed:
Issue #7277 - Updating based on review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented:
@sbordet changes applied

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,6 +301,70 @@ public EndPoint getEndPoint()
+    public void setLocalNameOverride(String overriddenLocalName)
+    {
+        this.overriddenLocalName = overriddenLocalName;
+    }
```
Not needed anymore.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,6 +301,70 @@ public EndPoint getEndPoint()
+        if (local != null)
+            return local.getHostString();
+
+        try
+        {
+            String name = InetAddress.getLocalHost().getHostName();
+            if (StringUtil.ALL_INTERFACES.equals(name))
+                return null;
+            return name;
+        }
+        catch (UnknownHostException e)
+        {
+            LOG.ignore(e);
+        }
```
Don't think this code should be there.
It's typically wrong if `/etc/hosts` is not setup properly, it does not work for Unix-Domain, leaks the internal host name or IP, etc.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1025,9 +1025,9 @@ public String getLocalName()
+                return formatAddrOrHost(localName);
         }
 
         try
```
See above, remove this code?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -298,6 +301,24 @@ public int getAcceptors()
+     * Optional override of connection local name used within application API layer
+     * when identifying the local host name of a connected endpoint.
+     */
+    public void setLocalName(String localName)
```
Do we just need a `localName` or we actually need a `localAuthority` (i.e. host _and_ port)?
@gregw thoughts?

gregw replied:
authority is better

joakime replied:
Still on AbstractConnector?
As in `AbstractConnector.setLocalAuthority(String authority)`?

joakime replied:
Changed to use authority instead.
Host AND port must be specified though.

joakime committed:
Issue #7277 - More updates based on review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Switch from setLocalName to setLocalAuthority per review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Fixing NPE from updated getLocalPort

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -97,6 +97,7 @@
     private MetaData.Response _committedMetaData;
     private RequestLog _requestLog;
     private long _oldIdleTimeout;
+    private String overriddenLocalName;
```
this is not used

joakime replied:
Removed

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -298,6 +302,42 @@ public int getAcceptors()
+     * @return Returns the optional local authority override
+     */
+    @ManagedAttribute("local authority")
+    public HostPort getLocalAuthority()
+    {
+        return _localAuthority;
+    }
```
This method should always return a local authority, either one set explicitly or one that is determined dynamically in `doStart` (or `open`).
Then the layers about can just ask for the authority and not have to check if null and then look at the address etc.

gregw replied:
Also make this a method on `NetworkConnector`

joakime replied:
This is not a NetworkConnector specific concept.
We have UnixDomainConnector and LocalConnector that can benefit from setLocalAuthority() and are not network connectors.

joakime replied:
We cannot "always return a local authority".
I tried it.
Going this route breaks expectations in ForwardRequestCustomizer and DetectorConnection and ProxyEndPoint behaviors, so I reverted it.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,6 +299,73 @@ public EndPoint getEndPoint()
+                return localAuthority.getHost();
+        }
+
+        InetSocketAddress local = getLocalAddress();
+        if (local != null)
+            return local.getHostString();
+
```
We should never fall through to here.  If there `Connector` is a `NetworkConnector` it should always provide a local authority, otherwise return null.

joakime replied:
See comment above about this feature not being NetworkConnector specific.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,6 +299,73 @@ public EndPoint getEndPoint()
+    {
+        Connector connector = getConnector();
+
+        if (connector instanceof AbstractConnector)
```
Ditto, test should be for `NetworkConnector` which should always provide the local port

gregw replied:
But also let's not convert to Host/Port here.  Might need to leave these as deprecated, but pass the local Authority all the way to the `Request`

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,6 +299,73 @@ public EndPoint getEndPoint()
+     * @return the local name if overridden, or the local address, or
+     * null in the case of no local address (usually seen in connectors not based on IP networking).
+     */
+    public String getLocalName()
```
Why not keep the `getLocalAuthority()` here and only convert  to host/port in `Request`

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1050,10 +1039,16 @@ public String getLocalName()
+            int localPort = _channel.getLocalPort();
+            if (localPort > 0)
+                return localPort;
+            InetSocketAddress local = _channel.getLocalAddress();
+            if (local != null)
+                return local.getPort();
```
We should not have to do this fall back yet again here.   Either is a local authority (so we have a port) or there is not (and we don't)

joakime replied:
Tried this.
Going this route breaks expectations in ForwardRequestCustomizer and DetectorConnection and ProxyEndPoint behaviors, so I reverted it.

joakime committed:
Issue #7277 - Updates based on review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Reverting predefined Local Authority

+ Breaks ...
  - DetectorConnectionTest
  - ProxyEndPoint
  - ForwardedRequestCustomizerTest

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Request.getServerName() must return non-null

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - LocalConnector should not have LocalAuthority default

+ Breaks ResponseTest.testSendRedirect() expectations

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Reverting changes unrelated to this PR

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Reverting changes unrelated to this PR

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Reverting whitespace changes unrelated to this PR

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented:
I can make all but the EmptyHost test pass on an unmodified jetty-9 with the following ammendment to the unit test:
```java
    private CloseableServer startServer(HostPort localAuthority) throws Exception
    {
        Server server = new Server();

        ServerConnector connector;
        if (localAuthority == null)
        {
            connector = new ServerConnector(server);
        }
        else
        {
            InetSocketAddress authorityInetSocketAddress = InetSocketAddress.createUnresolved(localAuthority.getHost(), localAuthority.getPort());
            connector = new ServerConnector(server)
            {
                @Override
                protected ChannelEndPoint newEndPoint(SocketChannel channel, ManagedSelector selectSet, SelectionKey key) throws IOException
                {
                    SocketChannelEndPoint endpoint = new SocketChannelEndPoint(channel, selectSet, key, getScheduler())
                    {
                        @Override
                        public InetSocketAddress getLocalAddress()
                        {
                            return authorityInetSocketAddress;
                        }
                    };
                    endpoint.setIdleTimeout(getIdleTimeout());
                    return endpoint;
                }
            };
        }
        connector.setPort(0);
```

So I'm really reluctant to make big changes in jetty-9 when a simple extention of `ServerConnector` server connector can achieve pretty much the same result.

joakime commented:
This PR replaced with PR #7316
